### PR TITLE
Add support for AD through sendrecv

### PR DESCRIPTION
**Pull request by PhilipVinc**
I investigated the matter, and it seems not possible to support send/recv in reverse mode.
Only in forward mode...

### CODE REVIEW

File: mpi4jax/_src/collective_ops/sendrecv.py [hunk 1]
```diff
@@ -234,10 +234,68 @@ def mpi_sendrecv_abstract_eval(
...
+
+    # swap the sender and receiver
+    res, token = mpi_sendrecv_p.bind(
+        out_tan,
+        out_tan,
+        token,
+        source=dest,
+        dest=source,
+        sendtag=sendtag,  # TODO: could maybe be smarter about send and receive tags...
+        recvtag=recvtag,  # TODO: could maybe be smarter about send and receive tags...
```
Comment (dionhaefner) [inline mpi4jax/_src/collective_ops/sendrecv.py:285]: I think the tags are fine. What's your issue with them?

File: tests/collective_ops/test_sendrecv.py [hunk 2]
```diff
@@ -104,3 +104,23 @@ def test_sendrecv_scalar_jit():
...
+def test_sendrecv_grad():
+    from mpi4jax import sendrecv
+
+    arr = jnp.ones((3, 2)) * (rank + 1)
+    _arr = arr.copy()
+
+    other = 1 - rank
+
+    def f(x):
+        x, token = sendrecv(x, x, source=other, dest=other)
```
Comment (dionhaefner) [inline tests/collective_ops/test_sendrecv.py:119]: Maybe have a test where `x` is also manipulated *before* calling `sendrecv`, so the transmitted gradient isn't just 1.

**Review by dionhaefner — COMMENTED** [review]
Looks promising! I just think it needs a little more testing (I find it hard to reason about AD behavior in JAX). Maybe also a test case like in #72.

File: mpi4jax/_src/collective_ops/sendrecv.py (hunk 1 above)
Comment (PhilipVinc) [reply mpi4jax/_src/collective_ops/sendrecv.py:285]: I'm unsure what tags are for, actually.
If they are to avoid deadlocks between messages sent/received or discriminating messages...?

Since this creates effectively another send_recv, maybe we should have send and recv tags for the tangents as inputs... (but it hsould be specified in the rule above, not here)

**Comment by PhilipVinc** [comment]
It's actually quite hard.
I find `jax.make_jaxpr(jax.grad(f))` quite enlightening.

File: mpi4jax/_src/collective_ops/sendrecv.py (hunk 1 above)
Comment (dionhaefner) [reply mpi4jax/_src/collective_ops/sendrecv.py:285]: I think tags are for situations like this:

```
if rank == 0:
    recv(..., source=1)
    recv(..., source=2)
elif rank == 1 or rank == 2:
    send(..., dest=0)
```

If the message from rank 2 arrives first, this would deadlock without tags. But with tags, it is able to use the right message.

Re-using the same tags as in the forward pass should be the correct thing to do.

Comment (PhilipVinc) [reply mpi4jax/_src/collective_ops/sendrecv.py:285]: Ok, i'll delete the comment

**Comment by dionhaefner** [comment]
Thanks, LGTM!
